[JS] Update Middleware Docs (#1234)
SAP/cloud-sdk: docs-js/features/middleware.mdx

## Example Hello World

A very simple middleware logs a 'Hello World' and leaves the original function unchanged.

```ts
const helloWorld: Middleware = function (options) {
  console.log('hello world');
  return options.fn;
};
```

Note that the log statement is executed when the middleware is added and not when the function executes.
If you want to trigger the log on function execution you have to create a new function in the middleware.

```ts
const helloWorld: Middleware = function (options) {
  return request => {
    console.log('hello world');
    options.fn(request);
  };
};
```

Note that a middleware is **not** actually executing the original function.
It adds functionality to the function and creates a new function from it.

In the example the custom code executes before the function `fn`.
This will result in a stack like behaviour for multiple middlewares.
For a queue like behaviour, add the code after the function resolves:

```ts
const helloWorld: Middleware = function (options) {
  return request => {
    options.fn(request).then(response => {
      console.log('hello world');
      return response;
    });

>   };

Author: Async await is not possible because then it would return a Promise<MiddlewareFunction>

};
```

## Example Multiple Middlewares

Often you will use multiple small middlewares to add functionality to the request.
Assume you want to include a `timeout` and `fallback` middleware.
The implementation of the middlewares would look like:

```ts
const timeout: Middleware = function (options) {
  return request => Promise.race([options.fn(request), timeoutPromise]);
};

const fallback: Middleware = function (options) {
  return request => {
    try {
      return options.fn(request);
    } catch (e) {
      //implement some fallback logic
    }
  };
};
```

> Also here the middleware does **not** execute the function but adds functionality.

Author: No it does not it returns a function which will execute it but it does not call it.

This makes a chain of middlewares different from a (asynchronous) pipeline where the methods are executed after each other.

Multiple middlewares are added one by one to the request.
The order is from left to right in the array.

```ts
executeHttpRequest(
  { url: 'http://example.com' },
  {
    middleware: [timeout, fallback],
    method: 'get'
  }
);
```

The initial function is a `GET` request to `example.com`.
The first middleware adds a timeout to this request leading a new function.
The second middleware adds a fallback to this new function.

The middlewares in this example are implemented like a stack.
The fallback was added last so it will execute first.
It acts on the request with the timeout.
The timeout was added first and counts only the time for the actual execution without the fallback.

If you would use:

```ts
executeHttpRequest(
  { url: 'http://example.com' },
  {
    middleware: [fallback, timeout],
    method: 'get'
  }
);
```

the timout would apply to the time for the original request plus additional time for a potential fallback call.
From this example you see that the order of middlewares is crucial.

> For a list of default resilience middlewares, a timeout should be added first, then a circuit breaker followed by a retry.

Reviewer: ```suggestion
The provided default resilience middlewares should be added in the following order:
1. timeout
2. circuit breaker
3. retry.
```
[req] We should introduce our provided middlewares first
[req] Make it easier to understand the order.

  }
);
```

The API for a middleware is as follows:

```ts
export type MiddlewareFunction = (
  request: HttpRequest
) => Promise<HttpResponse>;

export type Middleware = (options: MiddlewareOptions) => MiddlewareFunction;

export interface MiddlewareOptions {
  fn: MiddlewareFunction;
  context: HttpMiddlewareContext;
  skipNext: () => void;
}
```

Every middleware needs to return a `MiddlewareFunction` instance which has the same parameters and return type as the input function `fn()`.
Let's investigate the different properties of the `MiddlewareOptions` interface:

- The `fn()` function is the returned function from the previous middleware.
  For the first one it is the original HTTP request from the SAP Cloud SDK.
- The `context` provides information on the request like URL, headers and HTTP method.
- The `skipNext` will exit all following middlewares

Note that a middleware should **not** execute the function but construct and return a `MiddlewareFunction`.
This makes a chain of middlewares different from a pipeline where the functions are executed one after another.
This may seem abstract now, but the example in the next section will bring some concreteness to the topic.

> ## Example Hello World

Reviewer 2: [req] Now after I looked through the examples a bit, I would like to suggest to actually structure the article such that the examples are there to explain a topic more visually, but the topic itself is more in the center of it.

"Example Multiple Middlewares" basically explains the order of execution when there are multiple middlewares.
I think it would be more natural to have a section on "Order of Execution" (or so) and within this section use the example to better explain the topic.

  return request => {
    options.fn(request).then(response => {
      console.log('hello world');
      return response;
    });
  };
};
```

## Example Multiple Middlewares

A good practice is to define reusable, single-purpose middlewares.
You can add multiple of those middlewares to combine the effects.
Assume you want to modify your call to have a timeout and use a fallback.
The implementation of these middlewares could look like this:

```ts
const timeoutMiddleware = options => {
  return request => Promise.race([options.fn(request), timeoutPromise]);
};

const fallbackMiddleware = options => {
  return request => {
    try {
      return options.fn(request);
    } catch (e) {
      //implement some fallback logic
    }
  };
};
```

Multiple middlewares are added one-by-one to the request.
The order is from left to right in the array.

```ts
executeHttpRequest(
  { url: 'http://example.com' },
  {
    middleware: [timeoutMiddleware, fallbackMiddleware],
    method: 'get'
  }
);
```

The initial function is a `GET` request to `http://example.com`.
The first middleware creates a new function adding a timeout to the HTTP request (see _h<sub>1</sub>_ in the picture below).

> The second middleware adds a fallback to this new function leading _h<sub>2</sub>_ in the picture below.

Author: I added a picture to the text

The middlewares in this example are implemented like a stack.

> The fallback was added last, so it will execute first.

Author: The problem is it depends if it is a stack or queue on how you implement it.

It acts on the HTTP request with the timeout (_h<sub>1</sub>_ in the picture below).
The timeout counts only the time for the actual execution of the original function _f<sub>0</sub>_.

<ThemedImage
  alt="Middleware execution order"
  sources={{
    light: useBaseUrl('img/middleware_light.svg'),
    dark: useBaseUrl('img/middleware_dark.svg')
  }}
  className="center"
  width="100%"
/>

If you switched the order of the middlewares:

```ts
executeHttpRequest(
  { url: 'http://example.com' },
  {
    middleware: [fallback, timeout],
    method: 'get'
  }
);
```

> The timout would apply to the time for the original request plus additional time for a potential fallback call (see _g<sub>1</sub>_ and _g<sub>2</sub>_ in the picture above).

Reviewer 2: ```suggestion
The timeout would apply to the time for the original request plus additional time for a potential fallback call (see _g<sub>1</sub>_ and _g<sub>2</sub>_ in the picture above).
```

From this example, you see that the order of middlewares is crucial.

The provided default resilience middlewares should be added in the following order:

1. timeout
2. circuit breaker
3. retry.
   The SAP Cloud SDK provides default resilience middlewares so that you do not have to worry about the details in most cases.
   You can find detailed information on the resilience topic in a [dedicated documentation](../guides/resilience).

## Example Middleware Context

The previous examples did not use the middleware context.
However, the context provides useful information you can use in your middleware.

```ts

> export interface HttpMiddlewareContext {

Reviewer 2: [pp] As above a high level JS Doc, e.g. "Represents the Execution Context of a Middleware" or something, would help introduce this interface a bit.

> export interface HttpMiddlewareContext {

Reviewer 2: [q] Why is this called `HttpMiddlewareContext` - are there other contexts as well?

Author: yes this is the concrete Http case wher ethe middleware function is also HttpRequest->HttpResponse

  readonly tenantId: string;
  readonly uri: string;
  readonly jwt?: string;
  readonly destinationName?: string;
  fnArgument: ArgumentT;
}
```

> The fallback middleware could consider the uri:

Reviewer 2: ```suggestion
The fallback middleware could consider the URI:
```

```ts
const fallbackMiddleware = options => {
  return (request: HttpRequest) => {
    try {
      return options.fn(request);
    } catch (e) {
      if (options.context.uri === 'http://system-one.com') {
        //do something
      } else {
        //do something else
      }
    }
  };
};
```

Note that the `fnArgument` is not marked as `readonly`.
This indicates that changes to this object will change the argument for the actual execution.

> The SAP Cloud SDK uses the axios client which has a `timeout` property.

Reviewer 2: [q] Should it be axios or Axios?

Author: most of the time it is used like this

Author: it is also in line with the style guide: https://github.com/SAP/cloud-sdk/blob/main/STYLEGUIDE.md#use-ordinary-font-for-software-products

You could set a timeout on client level like this:

```ts
const clientTimeoutMiddleware = options => {
  options.fnArgument.timeout = 500;
  return options.fn;
};
```

## Benefits of Middlewares

You could implement many use cases of middlewares also directly.
You could implement a fallback using a global try/catch.

```ts
try {
  executeHttpRequest(destination, options);
} catch (e) {
  //implement fallback logic
}
```

However, this has two disadvantages:

- The first one is more of cosmetic nature.
  The middlewares encourage you to implement the logic in a separate function which can be reused in different places.
  Your actual business code stays clean.
  This becomes even more obvious for [typed clients](./odata/execute-request) which represent you business requirements more directly.
- The second one is about the position where you can include the code.
  With the middleware you can include logic directly to the HTTP layer.
  This is much more powerful than code on the outer layer.

> Assume you want to replace the HTTP client of the SAP Cloud SDK uses with a different one.

Reviewer 2: [req] something is wrong in this sentence.

Sometimes you want to adjust the way the SAP Cloud SDK executes the axios request.

The middleware adds a very flexible way to add custom functionality to the request execution.
You can add multiple middlewares to adjust requests to your needs.

```ts
executeHttpRequest(
  { url: 'http://example.com' },
  {
    middleware: [middlewareA, middlewareB,...],
    method: 'get'
  }
);
```

The API for a middleware is as follows:

```ts
/**

>  * Type defining the function passed from one middleware to the next one.

Reviewer 2: [pp] This is very much from the implementation point of view. If I put myself into the shoes of a user maybe the "passing" is not something I have on my radar. This is the function that middlewares return (and we internally pass it as `fn()` to the next one).

 */
export type MiddlewareFunction = (
  request: HttpRequest
) => Promise<HttpResponse>;

/**
 * The actual middleware type returning a function.
 */

> export type Middleware = (options: MiddlewareOptions) => MiddlewareFunction;

Reviewer 2: [pp] I'd put this first.

/**
 * The input options of the middleware.
 */
export interface MiddlewareOptions {
  fn: MiddlewareFunction;
  context: HttpMiddlewareContext;
  skipNext: () => void;
}
```

Every middleware takes the `MiddlewareOptions` as input and returns a `MiddlewareFunction` instance.
The `MiddlewareOptions` have the following properties:

- The `fn()` function is the returned function from the previous middleware.

>   For the last one it is the original HTTP request from the SAP Cloud SDK.

Reviewer 2: [req] Those two sentences are ambiguous, because in the first we use "previous" in a sense of order of execution. In the second we use "last one" in the order as it appears in the array (aka. "first one" in the order of execution).

>   Note that the type of the input and return function are the same.

Reviewer 2: [req/q] Is that true? The "input" is of type `MiddlewareOptions` the return(ed?) function is of type `MiddlewareFunction`. I don't really understand what is meant by this...

- The `context` provides information on the request like URL, headers and HTTP method.

> - The `skipNext` will exit all following middlewares

Reviewer 2: [req] Styleguide

```suggestion
- `skipNext()` will exit all following middlewares
```

Note that a middleware should **not** execute the function but construct and return a `MiddlewareFunction`.
You can think of middlewares as a composition of functions where each middleware composes a new function.
This makes a chain of middlewares different from a pipeline where the functions are executed one after another.
This may seem abstract now, but the examples in the next sections will bring some concreteness to the topic.

## Implementation of a Middleware

> In this section you learn how to implement a middleware based on a didactic example.

Reviewer 2: [pp] It's fine, but I just don't like these kinds of introductions ;) 
```suggestion
```

The following middleware logs a text and leaves the original function unchanged.

```ts

> const doSomethingBeforeMiddleware = options => {

Reviewer 2: [pp] General remark: Instead of `options` I would destructure the options to have the `fn` at hand. Reads better in my opinion.

  return requestConfig => {
    console.log('Implement something here.');
    options.fn(requestConfig);
  };
};
```

In the example the custom code executes before the function `fn()`.
Do this when you want to adjust things before the actual HTTP request is executed.

> If you want to process the result of the function your middleware implement it the following way

Reviewer 2: [req] Something is missing in this sentence.

```ts
const doSomethingAfterMiddleware = options => {
  return requestConfig => {
    options.fn(requestConfig).then(response => {
      console.log('Do some postprocessing.');
      return response;

>     });

Reviewer 2: [req] Shouldn't we use `await` as we would do according to the style guide (I think)?
